File: leetcode/412.py
"""
Runtime: 44 ms, faster than 51.47% of Python3 online submissions for Fizz Buzz.
Memory Usage: 15.1 MB, less than 21.74% of Python3 online submissions for Fizz Buzz.
"""


class Solution:
    def fizzBuzz(self, n: int) -> List[str]:
        ans = ["" for i in range(n)]
        for i in range(3, max(3, n + 1), 3):
            ans[i - 1] += "Fizz"
        for i in range(5, max(5, n + 1), 5):
            ans[i - 1] += "Buzz"
        for i in range(len(ans)):
            if not ans[i]:
                ans[i] = str(i + 1)

        return ans

File: leetcode/412.cpp
// Runtime: 21 ms, faster than 5.23% of C++ online submissions for Fizz Buzz.
// Memory Usage: 10.1 MB, less than 34.21% of C++ online submissions for Fizz Buzz.
// O(N) time, O(1) memory
class Solution {
 public:
  vector<string> fizzbuzz_;
  Solution() : fizzbuzz_(vector<string>(1e4, " ")) {
    for (int i = 1; i <= 1e4; ++i) {
      string ans = (i % 3 == 0 || i % 5 == 0) ? "" : to_string(i);
      if (i % 3 == 0) {
        ans = "Fizz";
      }
      if (i % 5 == 0) {
        ans += "Buzz";
      }
      fizzbuzz_[i - 1] = ans;
    }
  }

  vector<string> fizzBuzz(int n) {
    return vector<string>(fizzbuzz_.begin(), fizzbuzz_.begin() + n);
  }
};
